Context `merge` function

khieta opened the pull request:
## Description of changes

Adds two new methods to the public API:
* `Context::merge` as suggested in #1013 
* `Context::values` which gets the key/value pairs out of the context. This function seemed like the most natural way to implement `merge`, and could be useful in general.

Makes a breaking error API change (because I can't help myself):
* Renamed `DuplicateKeyInRecordLiteralError` to `DuplicateKeyError` and added a "context" field to allow more precise error messages. The reason for this change is just that I didn't like the error message "duplicate key _ in a record literal" for the `Context` functions.

## Issue #, if available

Resolves #1013

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [X] A change (breaking or otherwise) that only impacts unreleased or experimental code.

I confirm that this PR (choose one, and delete the other options):

- [X] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [X] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.

cdisselkoen commented on cedar-policy-core/src/ast/expr.rs:
```diff
@@ -1155,12 +1154,10 @@ impl<T: Clone> ExprBuilder<T> {
-    DuplicateKeyInRecordLiteral(
-        #[from] expression_construction_errors::DuplicateKeyInRecordLiteralError,
-    ),
+    DuplicateKeyInRecordLiteral(#[from] expression_construction_errors::DuplicateKeyError),
```
probably rename the variant here too while we're at it

cdisselkoen commented on cedar-policy-core/src/ast/expr.rs:
```diff
@@ -1169,23 +1166,30 @@ pub mod expression_construction_errors {
         }
+
+        /// Make a new error with an updated `context` field
+        pub fn set_context(self, context: &'static str) -> Self {
```
I'm not sure if this is general idiomaticity guideline or not, but I would expect `set_` to be a method that takes `&mut Self`, and `with_` to be a method that takes `self` and returns `Self`, like `Value::with_maybe_source_loc()` in Core.  Take or leave this advice

john-h-kastner-aws replied:
I wouldn't make this `pub`. `pub(crate)` probably?

cdisselkoen commented on cedar-policy-core/src/ast/request.rs:
```diff
@@ -309,28 +316,28 @@ impl Context {
-    // PANIC SAFETY: This is safe due to the invariant on `self.context`, `self.context` must always be a record
-    pub fn iter<'s>(&'s self) -> Option<Box<dyn Iterator<Item = (&SmolStr, PartialValue)> + 's>> {
+    // PANIC SAFETY: This is safe due to the invariant (ContextRecord) on `self.context`
+    pub fn values(self) -> Box<dyn Iterator<Item = (SmolStr, PartialValue)>> {
```
Should we call this `.into_iter()` instead?  Maybe even have it be an implementation of the `IntoIterator` trait?

khieta replied:
I'll try this out tomorrow

khieta replied:
Latest version adds an implementation of `IntoIterator`. I don't think we can provide an `.iter()` function that doesn't perform cloning atm. I think the best solution would be to address #541 by storing the context as a hash map instead of a record expression.

cdisselkoen replied:
Nicely done :)

Want to add a comment to #541 with your suggestion for implementation?

cdisselkoen commented on cedar-policy/src/api.rs:
```diff
@@ -3384,6 +3384,34 @@ impl Context {
+                    k.to_string(),
+                    match v {
+                        ast::PartialValue::Value(val) => RestrictedExpression(
+                            ast::RestrictedExpr::new_unchecked(ast::Expr::from(val)),
```
might be useful to have a `ast::RestrictedExpr::from(val)` function.  From this vantage point it's a little harder to see why `new_unchecked` is safe here

john-h-kastner-aws commented on cedar-policy-core/src/ast/request.rs:
```diff
@@ -309,28 +316,28 @@ impl Context {
-                    map.iter()
-                        .map(|(k, v)| (k, PartialValue::Residual(v.clone()))),
-                )),
-                ExprKind::Unknown(_) => None,
```
How is the `Unknown` case handled now? Kind of worried it might hit the `panic!("internal invariant violation: expected a record, got {kind:?}"),` case

khieta replied:
Great question! I believe the `Context::from_expr` function is used on all paths for building a context, and this function must return a record value or record residual due to the handing of records in `RestrictedEvaluator::partial_interpret_internal`. I updated the comment about the `ContextRecord` invariant to reflect this, but would appreciate if you would double check my reasoning.

john-h-kastner-aws replied:
Looked at the code closer and I agree that this match is safe given the invariant. 

#1026 introduced `from_partial_value_unchecked` which makes the invariant depend on those call sites. I'd be happier if that didn't need the `_unchecked` constructor, but I am convinced that the invariant is maintained. We already have #541 for future refactoring.

khieta replied:
Latest version adds additional comments to `from_partial_value_unchecked` to clarify 🙂

shaobo-he-aws replied:
Would it make sense to implement #541 in this PR if it's convenient?

john-h-kastner-aws replied:
I expect that change to be a bit annoying, so I'd rather see it in a follow up PR. Up to Kesha if wants to do it ofc.

khieta replied:
Yeah, that will be a somewhat involved change, so I'll leave it for future work.

aaronjeline commented on cedar-policy-core/src/entities/json/err.rs:
```diff
@@ -80,10 +80,10 @@ pub enum JsonDeserializationError {
     #[error(transparent)]
     #[diagnostic(transparent)]
-    DuplicateKeyInRecordLiteral(DuplicateKeyInRecordLiteral),
+    DuplicateKey(DuplicateKey),
```
This is a nicer name

cdisselkoen commented on cedar-policy/src/api.rs:
```diff
@@ -3387,30 +3387,50 @@ impl Context {
+                            RestrictedExpression(ast::RestrictedExpr::from(val))
+                        }
                         ast::PartialValue::Residual(exp) => {
                             RestrictedExpression(ast::RestrictedExpr::new_unchecked(exp))
```
maybe a short comment justifying the `_unchecked`, like you have elsewhere in this PR?

cdisselkoen reviewed (approved):
Looks great

shaobo-he-aws commented on cedar-policy/src/api.rs:
```diff
@@ -3384,6 +3384,54 @@ impl Context {
     }
+
+    /// Combine two [`Context`]s, returning an error if they contain overlapping keys
+    pub fn merge(self, context: Self) -> Result<Self, ContextCreationError> {
```
Sorry for being late. I was wondering if we can generalize the argument to `impl Iterator<SmolStr, ...>`, since `Context` implements the conversion.

john-h-kastner-aws reviewed (approved):
looks good!

khieta commented:
I believe I've addressed all comments now. Will wait until tomorrow to merge because I will need to make some downstream fixes (caused by the change `Context::iter` -> `Context::into_iter`)

khieta commented:
Fix for DRT failure is here: https://github.com/cedar-policy/cedar-spec/pull/376. Need a re-approval before I can merge this PR.
